Client/Frontend/Browser/TopTabsViewController.swift

```diff
@@ -328,44 +311,192 @@ extension TopTabsViewController: TabSelectionDelegate {
+
+        self.pendingUpdates.append(self.calculateDiffWith(oldTabs, to: newTabs, and: reloadTabs))
+        if self.isUpdating || self.pendingReloadData {
+            return
+        }
+
+        self.isUpdating = true
+        let updates = self.pendingUpdates
+        self.flushPendingChanges()
+        let update = tableUpdate(updates: updates)
+
+        performUpdateWithChanges(update) { (_) in
+            // This handles the edge case where, during the animation we've toggled private mode
+            // Because we dont have a proper way of knowing when this transition is about to happen we have to do this check here
+            print("number of items in datastore AFTER UPDATE \(self.tabsToDisplay.count)")
+            if  !self.tabsMatchDisplayGroup(newTabs.first, b: self.tabsToDisplay.first) || self.pendingReloadData {
+                self.reloadData()
+            } else if self.pendingUpdates.isEmpty && !self.isUpdating && !update.inserts.isEmpty {
+                self.scrollToCurrentTab()
+            }
+        }
+    }
+
+    func performUpdateWithChanges(update: tableUpdate, completion: (Bool)-> Void) {
+        //Speed up the animation a bit to make it feel snappier
+        let newUpdates = update.reloads.filter { self.tabsToDisplay.count  > $0.row }
+        UIView.animateWithDuration(0.1) {
+            self.collectionView.performBatchUpdates({
+                self.collectionView.deleteItemsAtIndexPaths(Array(update.deletes))
+                self.collectionView.reloadItemsAtIndexPaths(Array(newUpdates))
+                self.collectionView.insertItemsAtIndexPaths(Array(update.inserts))
+                self.isUpdating = false
+                }, completion: completion)
+        }
+    }
+
+    private func flushPendingChanges() {
+        _oldTabs.removeAll()
+        pendingUpdates.removeAll()
+    }
+
+    private func reloadData() {
+        assertIsMainThread("reloadData must only be called from main thread")
+
+        if self.isUpdating {
+            self.pendingReloadData = true
+            return
+        }
+
+        isUpdating = true
+        self.newTab.userInteractionEnabled = false
+        UIView.animateWithDuration(0.2, animations: {
+            self.collectionView.reloadData()
+            self.collectionView.collectionViewLayout.invalidateLayout()
+            self.collectionView.layoutIfNeeded()
+            self.scrollToCurrentTab(true, centerCell: true)
+        }) { (_) in
+            self.flushPendingChanges()
+            self.isUpdating = false
+            self.pendingReloadData = false
+            self.newTab.userInteractionEnabled = true
+        }
     }
 }
 
 extension TopTabsViewController: TabManagerDelegate {
+
+    // Because we don't know when we are about to transition to private mode
+    // check to make sure that the tab we are trying to add is being added to the right tab group
+    private func tabsMatchDisplayGroup(a: Tab?, b: Tab?) -> Bool {
```

Comment:
Sometimes tabA is nil which means if I do it this way I have to unwrap the bool. That makes things messier than I like.